lib/compress/zstd_compress.c

```diff
@@ -1199,42 +1200,61 @@ static size_t ZSTD_resetCCtx_usingCDict(ZSTD_CCtx* cctx,
                             U64 pledgedSrcSize,
                             ZSTD_buffered_policy_e zbuff)
 {
+    /* We have a choice between copying the dictionary context into the working
+     * context, or referencing the dictionary context from the working context
+     * in-place. We decide here which strategy to use. */
+    /* TODO: pick reasonable cut-off size, handle ZSTD_CONTENTSIZE_UNKNOWN */
```

Comment:
When we don't know the content size I think we should assume it is small since it is dictionary compression.

Reply:
My concern with that approach is that failing to copy the dict in for a large compression is potentially a very expensive mistake.

So I think in order to do that, we would need a mechanism to perform the copy if it turns out that the input is large.

Reply:
> potentially a very expensive mistake

How much expensive ?

Reply:
I thought about it and I take it back. Since the dictionary only applies to the first 128K block, the cost is bounded. A quick A/B test shows that an attached dict is like 5% slower at 128KB. So that might be fine.

I'll look into this more.

Reply:
Note that a dictionary is presumed valid as long as input `<= windowSize`,
so this can be a lot more than 128K.

That being said, it could also be that dictionary gains become really tiny, of not meaningless, beyond a certain input size. In which case, it could be an acceptable trade off to no longer look into the dictionary beyond a certain input size, irrespective of the `windowSize`.